internal/metrics/metrics.go

```diff
@@ -1,17 +1,224 @@
+	return metricsBackend
+}
+
+// newPrometheusMetrics reads the status of the database, calculates all of the values of the metrics,
+// registers these metrics to the prometheus registry, and returns the registry and the metrics.
+// The registry and metrics can be modified from this struct from anywhere in the codebase.
+func newPrometheusMetrics() *PrometheusMetrics {
+	m := &PrometheusMetrics{
+		registry:                       prometheus.NewRegistry(),
+		CertificateRequests:            certificateRequestsMetric(),
+		OutstandingCertificateRequests: outstandingCertificateRequestsMetric(),
+		Certificates:                   certificatesMetric(),
+		ExpiredCertificates:            expiredCertificatesMetric(),
+		CertificatesExpiringIn1Day:     certificatesExpiringIn1DayMetric(),
+		CertificatesExpiringIn7Days:    certificatesExpiringIn7DaysMetric(),
+		CertificatesExpiringIn30Days:   certificatesExpiringIn30DaysMetric(),
+		CertificatesExpiringIn90Days:   certificatesExpiringIn90DaysMetric(),
+
+		RequestsTotal:    requestsTotalMetric(),
+		RequestsDuration: requestDurationMetric(),
+	}
+	m.registry.MustRegister(m.CertificateRequests)
+	m.registry.MustRegister(m.OutstandingCertificateRequests)
+	m.registry.MustRegister(m.Certificates)
+	m.registry.MustRegister(m.ExpiredCertificates)
+	m.registry.MustRegister(m.CertificatesExpiringIn1Day)
+	m.registry.MustRegister(m.CertificatesExpiringIn7Days)
+	m.registry.MustRegister(m.CertificatesExpiringIn30Days)
+	m.registry.MustRegister(m.CertificatesExpiringIn90Days)
+
+	m.registry.MustRegister(m.RequestsTotal)
+	m.registry.MustRegister(m.RequestsDuration)
+
+	m.registry.MustRegister(collectors.NewGoCollector())
+	m.registry.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
+	return m
+}
+
+// generateMetrics receives the live list of csrs to calculate the most recent values for the metrics
+// defined for prometheus
+func (pm *PrometheusMetrics) generateMetrics(csrs []certdb.CertificateRequest) {
+	var csrCount int = len(csrs)
+	var outstandingCSRCount int
+	var certCount int
+	var expiredCertCount int
+	var expiringIn1DayCertCount int
+	var expiringIn7DaysCertCount int
+	var expiringIn30DaysCertCount int
+	var expiringIn90DaysCertCount int
+	for _, entry := range csrs {
+		if entry.Certificate == "" {
+			outstandingCSRCount += 1
+			continue
+		}
+		if entry.Certificate == "rejected" {
+			continue
+		}
+		certCount += 1
+		expiryDate := certificateExpiryDate(entry.Certificate)
+		daysRemaining := time.Until(expiryDate).Hours() / 24
+		if daysRemaining < 0 {
+			expiredCertCount += 1
+		} else {
+			if daysRemaining < 1 {
+				expiringIn1DayCertCount += 1
+			}
+			if daysRemaining < 7 {
+				expiringIn7DaysCertCount += 1
+			}
+			if daysRemaining < 30 {
+				expiringIn30DaysCertCount += 1
+			}
+			if daysRemaining < 90 {
+				expiringIn90DaysCertCount += 1
+			}
+		}
+	}
+	pm.CertificateRequests.Set(float64(csrCount))
```

Comment:
Why do we do those type conversions instead of setting `csrCount` to be a float64 from the start?

Reply:
I didn't even realize VS code automatically put type conversions here. Changed.